[ROCm] hipSPARSELt Integration

petrex opened the pull request:
This PR integrates the HIP Sparse Library (hipSparseLt) in PyTorch. 
**The is a cherrypick PR to support ROCM 6.2**

- Supported input and output types: int8, float16, bf16
- Supported compute types: int32, float32
- Supported memory layout : col major only

The following UTs are passing on rocm 6.2 with gfx942

- test_indices_backend_cusparselt_cuda
- test_mm_sparse_first_TN_dense_input_shape0_backend_cusparselt_cuda_bfloat16
- test_mm_sparse_first_TN_dense_input_shape0_backend_cusparselt_cuda_float16
- test_mm_sparse_first_TN_dense_input_shape0_backend_cusparselt_cuda_float32
- test_mm_sparse_first_TN_dense_input_shape0_backend_cusparselt_cuda_int8
- test_mm_sparse_first_TN_dense_input_shape1_backend_cusparselt_cuda_bfloat16
- test_mm_sparse_first_TN_dense_input_shape1_backend_cusparselt_cuda_float16
- test_mm_sparse_first_TN_dense_input_shape1_backend_cusparselt_cuda_float32
- test_mm_sparse_first_TN_dense_input_shape1_backend_cusparselt_cuda_int8
- test_mm_sparse_first_TN_dense_input_shape2_backend_cusparselt_cuda_bfloat16
- test_mm_sparse_first_TN_dense_input_shape2_backend_cusparselt_cuda_float16
- test_mm_sparse_first_TN_dense_input_shape2_backend_cusparselt_cuda_float32
- test_mm_sparse_first_TN_dense_input_shape2_backend_cusparselt_cuda_int8
- test_mm_sparse_second_NN_dense_input_shape0_backend_cusparselt_cuda_bfloat16
- test_mm_sparse_second_NN_dense_input_shape0_backend_cusparselt_cuda_float16
- test_mm_sparse_second_NN_dense_input_shape0_backend_cusparselt_cuda_float32
- test_mm_sparse_second_NN_dense_input_shape0_backend_cusparselt_cuda_int8
- test_mm_sparse_second_NN_dense_input_shape1_backend_cusparselt_cuda_bfloat16
- test_mm_sparse_second_NN_dense_input_shape1_backend_cusparselt_cuda_float16
- test_mm_sparse_second_NN_dense_input_shape1_backend_cusparselt_cuda_float32
- test_mm_sparse_second_NN_dense_input_shape1_backend_cusparselt_cuda_int8
- test_mm_sparse_second_NN_dense_input_shape2_backend_cusparselt_cuda_bfloat16
- test_mm_sparse_second_NN_dense_input_shape2_backend_cusparselt_cuda_float16
- test_mm_sparse_second_NN_dense_input_shape2_backend_cusparselt_cuda_float32
- test_mm_sparse_second_NN_dense_input_shape2_backend_cusparselt_cuda_int8
- test_to_sparse_semi_structured_backend_cusparselt_cuda_bfloat16
- test_to_sparse_semi_structured_backend_cusparselt_cuda_float16
- test_to_sparse_semi_structured_backend_cusparselt_cuda_float32
- test_to_sparse_semi_structured_backend_cusparselt_cuda_int8
- test_unsupported_dim_backend_cusparselt_cuda
- test_unsupported_dtype_backend_cusparselt_cuda_complex128
- test_unsupported_dtype_backend_cusparselt_cuda_complex64
- test_unsupported_dtype_backend_cusparselt_cuda_float32
- test_unsupported_dtype_backend_cusparselt_cuda_float64
- test_unsupported_dtype_backend_cusparselt_cuda_int16
- test_unsupported_dtype_backend_cusparselt_cuda_int32
- test_unsupported_dtype_backend_cusparselt_cuda_int64
- test_unsupported_dtype_backend_cusparselt_cuda_int8
- test_unsupported_dtype_backend_cusparselt_cuda_uint8
- test_unsupported_shape_backend_cusparselt_cuda_bfloat16
- test_unsupported_shape_backend_cusparselt_cuda_float16
- test_unsupported_shape_backend_cusparselt_cuda_float32
- test_unsupported_shape_backend_cusparselt_cuda_int8
- test_cslt_sparse_mm_search_cuda_bfloat16
- test_cslt_sparse_mm_search_cuda_float16
- test_cslt_sparse_mm_search_cuda_float32
- test_cslt_sparse_mm_search_cuda_int8

**Expect subsequent PRs to support future hipsparseLT(ROCm 6.3) features/fixes: 1. row major 2. unified hip enum 3. mixed data type**

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @naromero77amd @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal

Commit:
Update CUDA to HIP mappings for cusparseLt in torch/utils/hipify/cuda_to_hip_mappings.py

Commit:
Update CUDA to HIP mappings for cusparseLt in torch/utils/hipify/cuda_to_hip_mappings.py

Commit:
Update CUDA to HIP mappings for cusparseLt in torch/utils/hipify/cuda_to_hip_mappings.py

Commit:
Update CUDA to HIP mappings for cusparseLt in torch/utils/hipify/cuda_to_hip_mappings.py

Commit:
hipSparseLt support compute type HIPSPARSELT_COMPUTE_32F

Commit:
Add support for hipsparseLt compute types look-up

Commit:
Update CUDA to HIP mappings for cusparseLt compute types

Commit:
hipsparseLT does not support R_32F compute type

Commit:
add USE_ROCM && ROCM_VERSION

Commit:
add check for R32_F data type

Commit:
Update lookup table to retrieve the datatype

Commit:
linter

Commit:
fix hipspraseLtDatatype_t

Commit:
drop duplicate const

Commit:
lint

Commit:
fix typo

Commit:
fix depricate hipsparse order

Commit:
build env for hipsparseLt

pruthvistony commented:
Note: hipSparseLT is only supported on MI300.

- Need to condition to skip on other HW.
- Need to skip the UTs related to cuSPARSELT/hipSparseLT on non-MI300 HW.

Commit:
fix include header

Commit:
add check hipsparseLT support arch

Commit:
update hipsparselt support matrix

Commit:
add device arch support guard

Commit:
include header fix

Commit:
cache the result; reduce device query

Commit:
build hipsparselt by default

petrex commented:
verify that the ci image:  ghcr.io/pytorch/ci-image:pytorch-linux-focal-rocm-n-py3-1b6850753b4c39818abf536373ed7c0ebda61f27
does not contain hipsparselt

Commit:
Add support for hipsparselt package if ROCM version is 6.1.0 or greater

Commit:
Fix duplicate call to find_package_and_print_version(hipsparselt) in LoadHIP.cmake

Commit:
use compressed_A.device().index() for checking hipSparseLtArch support

Commit:
Refactor hipSparseLtArch support check in cuSPARSELtOps.cpp

Commit:
fix computeType mapping/lib dependency/refactor

petrex commented:
@pytorchbot rebase

Commit:
disable hipsparseLt check for cuda

petrex commented:
shard 3, logs -> 100% passing but workflow failing. Not relevant.
2024-05-03T06:12:11.4874701Z test_scatter_gather_ops.py::TestScatterGatherCUDA::test_scatter_reduce_sum_cuda_uint8 PASSED [0.3310s] [100%]

petrex commented:
lintrunner job fail --> not relevant
```
 Cloning into '/home/ec2-user/actions-runner/_work/pytorch/pytorch/pytorch/pytorch/third_party/eigen'...
  Error: error: RPC failed; curl 56 Recv failure: Connection reset by peer
  Error: error: 4278 bytes of body are still expected
  fetch-pack: unexpected disconnect while reading sideband packet
  Error: fatal: early EOF
  Error: fatal: fetch-pack: invalid index-pack output
  Error: fatal: clone of 'https://gitlab.com/libeigen/eigen.git' into submodule path '/home/ec2-user/actions-runner/_work/pytorch/pytorch/pytorch/pytorch/third_party/eigen' failed
```

petrex commented:
UT shard 3 -> tests passing 100% but workflow fails. Not relevant.

```
2024-05-03T06:12:11.3879760Z test_linalg.py::TestLinalgCUDA::test_vector_norm_reduce_over_1D_vector_cuda_float64 PASSED [0.0097s] [100%]
2024-05-03T06:12:11.3879764Z 
2024-05-03T06:12:11.3880354Z - generated xml file: /v
```ar/lib/jenkins/pytorch/test/test-reports/python-pytest/test_linalg/test_linalg-b691a7d724704feb.xml -

petrex commented:
cuda build 

```
/var/lib/jenkins/workspace/aten/src/ATen/native/sparse/cuda/cuSPARSELtOps.cpp:165:10: error: unused variable ‘ishipSparseLt’ [-Werror=unused-variable]
  165 |     auto ishipSparseLt = false;
      |          ^~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
```

Commit:
drop ishipSparseLt from cuda build

Commit:
Refactor hipSparseLtArch support check in cuSPARSELtOps.cpp

Commit:
fix typo

petrex commented:
irrelevant test failure, shard 1/2

`2024-05-06T20:03:31.6787180Z test_linalg 2/3 failed!`

petrex commented:
@pytorchbot rebase

petrex commented:
looks like a false alarm

```
2024-05-06T23:34:51.6818749Z test_linalg.py::TestLinalgCUDA::test_vector_norm_reduce_over_1D_vector_cuda_float64 PASSED [0.0099s] [100%]
2024-05-06T23:34:51.6818755Z 
2024-05-06T23:34:51.6819330Z - generated xml file: /var/lib/jenkins/pytorch/test/test-reports/python-pytest/test_linalg/test_linalg-293106446d0ec4e4.xml -
2024-05-06T23:34:51.6819535Z ================= 224 passed, 103 skipped in 321.09s (0:05:21) =================
2024-05-06T23:34:51.6819658Z double free or corruption (!prev)
2024-05-06T23:34:51.6819799Z Got exit code -6 (SIGIOT)
2024-05-06T23:34:51.6819914Z Stopping at first consistent failure
2024-05-06T23:34:51.6820108Z The following tests failed consistently: ['ul']
2024-05-06T23:34:51.6820113Z 
2024-05-06T23:34:51.6820532Z FINISHED PRINTING LOG FILE of test_linalg 1/3 (test/test-reports/test_linalg_1.3_08c075c2f25a218c_.log)
2024-05-06T23:34:51.6820539Z 
2024-05-06T23:34:51.6820650Z test_linalg 1/3 failed!
2024-05-06T23:34:51.6820879Z Running test_quantization 4/10 ... [2024-05-06 23:34:51.157119]
2024-05-06T23:34:51.6821917Z Executing ['/opt/conda/envs/py_3.8/bin/python', '-bb', 'test_quantization.py', '-m', 'not serial', '--shard-id=4', '--num-shards=10', '-v', '-vv', '-rfEX', '-p', 'no:xdist', '--use-pytest', '-x', '--reruns=2', '--import-slow-tests', '--import-disabled-tests'] ... [2024-05-06 23:34:51.157474]
2024-05-06T23:34:51.6822039Z Emiting td_test_failure_stats_v2
2024-05-06T23:34:53.3309879Z /var/lib/jenkins/pytorch/tools/stats/upload_metrics.py:171: UserWarning: Error uploading metric td_test_failure_stats_v2 to DynamoDB: Unable to locate credentials
2024-05-06T23:34:53.3311408Z   warn(f"Error uploading metric {metric_name} to DynamoDB: {e}")
2024-05-06T23:34:53.3312073Z test_linalg 1/3 failed!
```

Commit:
add mi300x check, cusparselt/hipsparselt backend

Commit:
hipsparselt: HIP backend only support HIPSPARSE_ORDER_COL

Commit:
unskip tests for MI300x

Commit:
drop float32 from hipsparselt inference type

petrex commented:
@pytorchbot rebase

Commit:
fix leading dimension

Commit:
Seperate result buffer for hipSparseLtMatMul()

Commit:
skip tests that are not supported in ROCm atm

petrex commented:
@pytorchbot rebase

petrex commented:
irrelevant tests failure form CI...

[rocm / linux-focal-rocm6.1-py3.8 / test (default, 1, 6, linux.rocm.gpu.2)](https://hud.pytorch.org/pr/pytorch/pytorch/124320#25856602901) ([gh](https://github.com/pytorch/pytorch/actions/runs/9388960761/job/25856602901))
'ul'
[rocm / linux-focal-rocm6.1-py3.8 / test (default, 2, 6, linux.rocm.gpu.2)](https://hud.pytorch.org/pr/pytorch/pytorch/124320#25856603200) ([gh](https://github.com/pytorch/pytorch/actions/runs/9388960761/job/25856603200))
inductor/test_compiled_optimizers.py::CompiledOptimizerParityTestsCUDA::test_correctness_Adamax_use_closure_True_cuda_float32
[rocm / linux-focal-rocm6.1-py3.8 / test (default, 3, 6, linux.rocm.gpu.2)](https://hud.pytorch.org/pr/pytorch/pytorch/124320#25856603421) ([gh](https://github.com/pytorch/pytorch/actions/runs/9388960761/job/25856603421))
'ul'

petrex commented:
@pytorchbot rebase

petrex commented:
failing CI tests looks irrelevant

[rocm / linux-focal-rocm6.1-py3.8 / test (default, 1, 6, linux.rocm.gpu.2)](https://hud.pytorch.org/pr/pytorch/pytorch/124320#25866615325) ([gh](https://github.com/pytorch/pytorch/actions/runs/9392196441/job/25866615325))
'ul'
[rocm / linux-focal-rocm6.1-py3.8 / test (default, 2, 6, linux.rocm.gpu.2)](https://hud.pytorch.org/pr/pytorch/pytorch/124320#25866615492) ([gh](https://github.com/pytorch/pytorch/actions/runs/9392196441/job/25866615492))
test_foreach.py::TestForeachCUDA::test_unary_op_tensors_on_different_devices__foreach_tan_cuda_int16
[rocm / linux-focal-rocm6.1-py3.8 / test (default, 3, 6, linux.rocm.gpu.2)](https://hud.pytorch.org/pr/pytorch/pytorch/124320#25866615643) ([gh](https://github.com/pytorch/pytorch/actions/runs/9392196441/job/25866615643))
'ul'

pruthvistony commented:
@pytorchbot rebase

petrex commented:
@pytorchbot rebase

Commit:
stress CI

remove duplicate mapping

petrex commented:
@pytorchbot rebase

petrex commented:
@pytorchbot rebase

Commit:
update to support row_major_order and unified hip enum

update support check for hipSparseLt

cusparseLtDenseDescriptorInit update for hipsparselt

Commit:
fix cusparselt/hipsparselt enable

Commit:
lint

petrex commented:
@pytorchbot rebase

Commit:
fix typo function name

Commit:
Revert "update to support row_major_order and unified hip enum" for rocm 6.2 CI

This reverts commit e3c7b0b940e8ce4f2a9fd436972d1801841df595.

petrex commented:
@pytorchbot rebase

petrex commented:
@pytorchmergebot rebase

Commit:
drop rocm specific result pointer (update for rocm 6.3+)

petrex commented:
@pytorchbot rebase

Commit:
enable hipsparselt build

Commit:
update support arch check

Commit:
add arch check to _cslt_compress

Commit:
fix merge error

Commit:
drop sepeperate res_out

Commit:
hipsparselt is required for rocm 6.2+

Commit:
fix merge conflict/lint

Commit:
add specific rocm version check for sparse_mm()

Commit:
fix merge error/lint

Commit:
fix test : training tests only run on SM80

Commit:
fix C_type for hipsparselt

petrex commented:
@pytorchbot rebase

Commit:
Revert "fix merge error"

This reverts commit d4807e0bc534d8171de48f4164d2853f01433955.

Commit:
fix gfx942 import

petrex commented:
@pytorchbot rebase

petrex committed:
Merge branch 'main' into hipSPARSELt_rocm

Commit:
fix GFX_942 import

jcaip commented:
Hey @petrex thanks for working on this. I added 2:4 sparsity support to pytorch, so really interested in adding AMD support. Please let me know if there's anything I can do to help :)

petrex commented:
thanks @jcaip lets chat offline?
